### PR TITLE
Documentations for Google Summer of Code

### PR DESCRIPTION
**Pull request by PowerfooI**
## Summary

1. GSoC ideas list
2. Contributor guidance
3. Develop ob-operator locally
4. Deploy ob-operator and OceanBase cluster locally

### CODE REVIEW

File: docs/en_US/gsoc-ideas.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# GSoC 2024 Ideas List
+
+Hi there! This is the ideas list of OceanBase for Google Summer of Code 2024! 
+
+As a first-year mentor organization, we focus the development ideas on ob-operator, a kubernetes operator which helps deploy and manage OceanBase cluster on kubernetes cluster seamlessly. There are four projects for contributors, and we'll offer delicate guidance for every choice.
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:5]: delicate or dedicated ?

File: docs/en_US/gsoc-ideas.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+Enjoy your summer of code!
+
+## 1. CLI tool
+
+* Project Description: The primary method to manage CRDs of ob-operator is manipulating YAML manifests which is not easy enough for new-coming users to get started. Within the CLI tool, features like component installation, demo setup,  resource management and necessary validation can be implemented.
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:11]: "Within the CLI tool, features like component installation, demo setup,  resource management and necessary validation can be implemented" this is a little bit strange
how about change it to 
"We'd like to provide features such as component installation, demo setup,  resource management and necessary validation within a CLI tool"

Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:11]: CRDs of ob-operator or oceanbase ?

File: docs/en_US/gsoc-ideas.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+* Required Skills: Golang, Kubernetes
+* Project Size: medium/large
+* Expected Outcomes: A complete CLI tool to control and manage CRDs in ob-operator, in other words, to manage clusters, tenants, backups on ob-operator in kubernetes cluster. Or a module in the CLI tool.
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:14]: we need to make it concrete, list the minimal features set, and other extra features as a bonus

File: docs/en_US/gsoc-ideas.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+* References: 
+  * [kubernetes/client-go](https://github.com/kubernetes/client-go)
+  * [spf13/cobra](https://github.com/spf13/cobra)
+  * [manifoldco/promptui](https://github.com/manifoldco/promptui)
+
+
+## 2. Light-weighted Operations
+
+* Project Description: Use Light-weighted operation task types to influence the status of resources instead of modifying specifications of those bigger resources.
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:23]: the operations also need to modify specifications of the resource

File: docs/en_US/gsoc-ideas.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+* Required Skills: Golang, Development of Kubernetes controller, Docker
+* Project Size: medium
+* Expected Outcomes: Implementation of CRD(s) and corresponding controller(s) for cluster and tenant, which can trigger small operational actions and reveal the progress through `status` fields.
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:26]: 1. "for cluster and tenant" need to be changed to "for obcluster and obtenant operations"
2. need to support all the operations ob-operator provides currently

File: docs/en_US/gsoc-ideas.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+* References: 
+  * [Tenant operation: Failover](https://en.oceanbase.com/docs/common-oceanbase-database-10000000001106036)
+  * [Tenant operation: Replay log](https://en.oceanbase.com/docs/common-oceanbase-database-10000000001103949)
+  * [Architecture of ob-operator](https://oceanbase.github.io/ob-operator/docs/en_US/arch.html)
+
+
+## 3. Alertmanager Integration
+
+* Project Description: There are some features about cluster management in OceanBase dashboard, a web-based dashboard application, and tenant management features are coming soon. OceanBase dashboard has bundled prometheus to manage time series metrics data by now. It's recommended that Alertmanager performs alerting tasks along with promethues. Integrating Alertmanager into OceanBase dashboard could be an attracive feature.
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:35]: just give a general description of oceanbase dashboard, there's no need to mention the upcoming feature

"It's recommended that Alertmanager performs alerting tasks along with promethues" what does this mean?

File: docs/en_US/gsoc-ideas.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+* Required Skills: Web development(React), Kubernetes, Prometheus, Alertmanager
+* Project Size: medium
+* Expected Outcomes: A functional panel in OceanBase dashboard in which users could view alert events, configure alert rules, and define alert templates in the scope of Alertmanager.
+* References: 
+  * [Alertmanager](https://prometheus.io/docs/alerting/latest/alertmanager/)
+  * [prometheus/alertmanager](https://github.com/prometheus/alertmanager)
+  * [OceanBase - Monitor - Overview](https://en.oceanbase.com/docs/common-oceanbase-database-10000000001103563)
+
+
+## 4. OceanBase Database Proxy integration
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:45]: it's not integration, it's a feature in oceanbase dashboard
how about "ODP management in OceanBase Dashboard"

File: docs/en_US/gsoc-ideas.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+* Project Description: OceanBase Database Proxy (ODP), also called OBProxy, is a dedicated proxy server for OceanBase Database. Core features of ODP include connection management, optimal routing, high-performance forwarding, easy O&M, high availability, and proprietary protocol. ODP should be integrated into OceanBase dashboard to enhance proxy management.
+* Required Skills: Web development(React), Golang Kubernetes,
+* Project Size: medium
+* Expected Outcomes: A functional panel in OceanBase dashboard where users could setup, configure and even delete ODP for specific OceanBase cluster.
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:50]: is it an independent module or under obcluster

File: docs/en_US/gsoc-ideas.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+* References: 
+  * [OceanBase Database Proxy](https://en.oceanbase.com/docs/odp-en)
+
+
+## 5. Accounts management and RBAC admission control
+
+* Project Description: OceanBase Dashboard is a web-based management application for ob-operator, with support for managing cluster and tenant clearly. Currently, it has a quite simple account system that stores and retrieves user credentials with `Secret` resource. And, it lacks any form of admission control. So it would be a good starting point to develop an account management module, complemented by a robust RBAC (Role-Based Access Control) permissions system.
```
Comment (chris-sun-star) [inline docs/en_US/gsoc-ideas.md:57]: application -> platform

File: docs/en_US/contributor-guidance.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
+# Contributor Guidance for Our Project
+
+Welcome and thank you for your interest in contributing to our project! We are excited to have you join our community and look forward to your valuable contributions. This document serves as a guide to help you get started with contributing to our project in a way that is efficient and aligns with our community principles.
+
+## Introduction to the Project
```
Comment (chris-sun-star) [inline docs/en_US/contributor-guidance.md:5]: delete line 5 ~ 7

File: docs/en_US/contributor-guidance.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+## How to Contribute
+
+Here is a brief overview of how you can contribute to our project:
+
+1. Familiarize Yourself: Get to know the project by reading the documentation, understanding the codebase, and identifying the areas you are interested in.
+
+2. Find an Issue: Look at our issue tracker to find tasks that interest you. We also welcome contributions that improve the project in ways not yet identified in our tracker.
+
+3. Communicate: Before starting to work on an issue, let the community know to avoid duplicate efforts and to get any necessary input or guidance.
```
Comment (chris-sun-star) [inline docs/en_US/contributor-guidance.md:33]: starting to -> starting

File: docs/en_US/contributor-guidance.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+4. Fork and Clone: Fork the repository and clone it to your local machine. This will be your private workspace.
+
+5. Make Changes: Implement your changes in your local environment adhering to our development practices and coding standards.
+
+6. Write Tests: Ensure your changes are tested. We strive to maintain a high level of code quality and reliability.
+
+7. Document Your Work: Update existing documentation or create new documentation as needed. Good documentation is just as important as good code.
```
Comment (chris-sun-star) [inline docs/en_US/contributor-guidance.md:41]: change "documentation" to "documents"

File: docs/en_US/contributor-guidance.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+8. Create a Pull Request: Once your changes are ready, submit a pull request. Provide a clear description of the problem and solution, including any relevant issue numbers.
```
Comment (chris-sun-star) [inline docs/en_US/contributor-guidance.md:43]: issue numbers -> issue

File: docs/en_US/contributor-guidance.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+9. Code Review: Engage with the code review process, addressing any feedback provided by the project maintainers.
```
Comment (chris-sun-star) [inline docs/en_US/contributor-guidance.md:45]: Code Review is for maintainers, for contributors, this should be "Modify your code"

File: docs/en_US/develop-locally.md [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+## Background
+
+ob-operator depends on [kubebuilder](https://kubebuilder.io/introduction), an operator framework maintained by kubernetes SIGS. It offers convenient utilities to bootstrap an operator and manage API types in it. Like other operator frameworks, kubebuilder depends on kubernetes [controller runtime](https://github.com/kubernetes-sigs/controller-runtime) either, which is an excellent reference to know how kubernetes dispatch events and reconcile resources.
+
+## Disable Webhook and CertManager
+
+There are many configuration items that marked by `[CERTMANAGER]` and `[WEBHOOK]` in the two files `config/crd/kustomization.yaml` and `config/default/kustomization.yaml`. The are used to enable and configure webhooks in real kubernetes deployment. Because we want to run controller manager locally, we need to disable them.
+
+You could just apply the latest `deploy/operator.yaml` manifest and delete the following resources to deploy CRDs and make controller manager uninstalled.
```
[the remark this answers is not in the document]
Comment (PowerfooI) [reply docs/en_US/develop-locally.md:13]: It means that developers should install CRDs and delete controller manager

File: docs/en_US/deploy-locally.md [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+
+## Hardware requirements
+
+* CPU: 4 Core available
+* Memory: 10 GB available (A Single-node OceanBase cluster takes 8 ~ 9 GB memory)
+* Disk space: 120 GB available
+
+## Software requirements
+
+You should install following two software before we start. Installation instructions could be found on their sites. You can get minkube by brew install minikube if you develop on macOS.
```
Comment (chris-sun-star) [inline docs/en_US/deploy-locally.md:11]: how about change this line to "You need to install the following softwares before getting started. Installation instructions can be found on their websites."

File: docs/en_US/deploy-locally.md [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+* [kubernetes/minikube](https://github.com/kubernetes/minikube): Run kubernetes locally
+* [Docker desktop](https://www.docker.com/products/docker-desktop/): Manage container development locally
```
Comment (chris-sun-star) [inline docs/en_US/deploy-locally.md:13]: development -> deployments ?

File: docs/en_US/deploy-locally.md [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+
+In order to make sure that the hardware limit are enforced to VM in minikube, we set configurations of minikube again by:
+
+```shell
+minikube config set cpus 4
+minikube config set disk-size 120GB
+minikube config set memory 10GB
+```
+
+Here we go! Just type `minikube start` and enter, we could see it starts fetching necessary images. After it says initialization is finished, we could type `kubectl get pods -A` (or `minikube kubectl -- get pods -A`) to see all pods in minikube cluster.
```
Comment (chris-sun-star) [inline docs/en_US/deploy-locally.md:34]: could -> can

File: docs/en_US/deploy-locally.md [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+
+Tips: Perform `minikube dashboard` to open kubernetes dashboard, everything in the cluster could be more clear.
+
+### 3. Install cert-manager
```
Comment (chris-sun-star) [inline docs/en_US/deploy-locally.md:38]: the process to deploy the components in kubernetes cluster is duplicated
I'd like to suggest to change this document only focused on setting up the kubernetes environment

File: docs/en_US/contributor-guidance.md (hunk 11 above)
Comment (PowerfooI) [reply docs/en_US/contributor-guidance.md:33]: It should be `start to do` or `start doing`. Maybe `start to` is a more proper choice.

File: docs/en_US/deploy-locally.md (hunk 17 above)
Comment (PowerfooI) [reply docs/en_US/deploy-locally.md:13]: It's not a typo. What I mean is writing dockerfile, building and running images, manage manifests and so on. Not only deployments.

File: docs/en_US/deploy-locally.md (hunk 19 above)
Comment (PowerfooI) [reply docs/en_US/deploy-locally.md:38]: The existing `deploy` documentation reveals only installation instructions of ob-operator. No clusters and tenants are involved, while they are essential for developers to see what they would behave.

File: docs/en_US/develop-locally.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (chris-sun-star) [reply docs/en_US/develop-locally.md:13]: we can provide a make command  to just generate the crds

**Review by chris-sun-star — APPROVED** [review]

**Pull request merged**